drivers: video: introduce VIDEO_FOURCC_TO_STR

josuah opened the pull request:
This adds a macro to generate a C99 compound literal string, which allow to use it in debug log messages, such as:
`LOG_DBG("The pixel format is '%s'", VIDEO_FOURCC_STR(fmt->pixelformat));`

The [previous proposal](https://github.com/zephyrproject-rtos/zephyr/pull/79476) was more awkward: `printf("video format: " PRIvfmt, PRIvfmt_arg(&fmt));`

josuah committed:
drivers: video: introduce VIDEO_FOURCC_TO_STR

This adds a macro to generate a C99 compound literal string, which allow
to use it in debug log messages, such as:
LOG_DBG("The pixel format is '%s'", VIDEO_FOURCC_STR(fmt->pixelformat));

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Wrapping it in parenthesis seemed necessary in some cases.

kartben commented:
> Wrapping it in parenthesis seemed necessary in some cases.

Yep, totally the right approach FWIW :) https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#:~:text=Expressions%20resulting%20from%20the%20expansion%20of%20macro%20parameters%20shall%20be%20enclosed%20in%20parentheses

kartben commented on include/zephyr/drivers/video.h:
```diff
@@ -833,6 +833,17 @@ void video_closest_frmival(const struct device *dev, enum video_endpoint_id ep,
+ * @param fcc 32-bit integer to be converted
+ * @return Four-character-string built out of this fourcc.
+ */
+#define VIDEO_FOURCC_TO_STR(u) ((char []){(u) >> 0, (u) >> 8, (u) >> 16, (u) >> 24, '\0'})
```
don't each byte need to explicitly be cast to char? probably warning-prone otherwise?
also not sure about the naming `u`, and I think you might face endianness issues?

```c
#define VIDEO_FOURCC_TO_STR(fourcc) ((char []){ \
    (char)((fourcc) & 0xFF), \
    (char)(((fourcc) >> 8) & 0xFF), \
    (char)(((fourcc) >> 16) & 0xFF), \
    (char)(((fourcc) >> 24) & 0xFF), \
    '\0' \
})
```

josuah replied:
> you might face endianness issues

The `pixelformat` used everywhere in `video.h` is what FOURCC are used for, and is always in the native endianess of the CPU. That can lead to surprises if logging something directly from the network, I added a comment.

Thank you for the improvement and reminding that clarity and correctness are better than compactness.

josuah replied:
- little/big-endian (LSB/MSB first)
- native-endian (what the CPU has)
- confused-endian (the opposite) [like FOURCCs](https://en.wikipedia.org/wiki/FourCC#Compiler_support)

ngphibang commented:
I think the PR should also include changes that uses this macro, at least in the video samples. This avoids adding additional PRs later for it.

ngphibang reviewed (changes requested):
I think the PR should also include changes that uses this macro, at least in the video samples. This avoids adding additional PRs later for it (and avoid adding things that has no usage). Otherwise, LGTM

josuah committed:
samples: drivers: video: use a macro to simplify format logging

A macro introduced allows to log four-character-codes (FOURCC) as one
string instead of enumerating each individual character. This saves a bit
of room in the code and is less error-prone.

Signed-off-by: Josuah Demangeon <me@josuah.net>
